SPIFlashLogger.device.lib.nut

```diff
@@ -0,0 +1,847 @@
+            ++i;
+        } while (i <= (SPIFLASHLOGGER_SECTOR_SIZE / SPIFLASHLOGGER_CHUNK_SIZE));
+        // there is no free free chunk, all sector is busy
+        return SPIFLASHLOGGER_SECTOR_SIZE;
+    }
+
+    function isFree() {
+        return this._status == 0xFF;
+    }
+
+    function erase() {
+        // Erase sector
+        local sec = (this._start - _logger._start) / SPIFLASHLOGGER_SECTOR_SIZE;
+        // erase current sector
+        _logger._erase(sec, sec + 1);
+
+        this._status = 0xFF;
+        this._chunkMap = 0xFF;
+        this._pos_write = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+        this._pos_read = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+
+        _eraseBeforeWrite = false;
+    }
+
+    function setPosition(pos) {
+        if ((math.abs(pos - _pos_write) > SPIFLASHLOGGER_CHUNK_SIZE) ||
+            (pos <= SPIFLASHLOGGER_SECTOR_METADATA_SIZE && !this.isFree() ||
+                this.isStoppedWriting())) {
+            // Wrong position
+            _eraseBeforeWrite = true;
+        } else
+            _pos_write = pos;
+    }
+};
+
+class SPIFlashLogger.LoggerSerializedObject {
+    _flash = null;
+    _logger = null;
+    _addr = null;
+    _isValid = false;
+    _payload = null;
+    _len = 0;
+
+
+_debug = false;
+    constructor(addr, logger, flash) {
+        _logger = logger;
+        _flash = flash;
+        _addr = addr;
+
+        _isValid = false;
+        _payload = null;
+        _len = 0;
+
+        init(addr);
+    }
+
+    function init(pos) {
+        local requested_pos = pos;
+        _logger._enable();
+        // Get the meta (for checking) and the object length (to know how much to read)
+        local marker = _flash.read(pos, SPIFLASHLOGGER_OBJECT_MARKER_SIZE).tostring();
+        local len = _flash.read(pos + SPIFLASHLOGGER_OBJECT_MARKER_SIZE, 2).readn('w');
+        _logger._disable();
+        _len = len;
+        _isValid = (marker == SPIFLASHLOGGER_OBJECT_MARKER && len > 0 && len < _logger._size);
+    }
+
+    function isValid() {
+        return _isValid;
+    }
+
+    function getAddress() {
+        return _addr;
+    }
+
+    function getPayload() {
+        if (!_isValid)
+            return null;
+
+        if (_payload != null)
+            return _payload;
+
+        local pos = _addr;
+        local serialised = blob(SPIFLASHLOGGER_OBJECT_HDR_SIZE + _len);
+        local leftInObject;
+        _logger._enable();
+        // while there is more object left, read as much as we can from each sector into `serialised`
+        while (leftInObject = serialised.len() - serialised.tell()) {
+            // Decide what to read
+            local sectorStart = pos - (pos % SPIFLASHLOGGER_SECTOR_SIZE);
+            local sectorEnd = sectorStart + SPIFLASHLOGGER_SECTOR_SIZE;
+            local leftInSector = sectorEnd - pos;
+
+            // Read it
+            local read;
+            if (leftInObject < leftInSector) {
```

Comment:
This can be simplified like this:

```
    local left = leftInObject < leftInSector ? leftInObject : leftInSector;
    read = _flash.read(pos, left;
    assert(read.len() == left);
```

Reply:
good point, agree (just copy past from original code)